ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/MyAppPurchaseLogic.kt

```diff
@@ -0,0 +1,118 @@
+import com.revenuecat.purchases.CustomerInfo
+import com.revenuecat.purchases.Package
+import com.revenuecat.purchases.PresentedOfferingContext
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.models.GoogleReplacementMode
+import com.revenuecat.purchases.models.PurchaseState
+import com.revenuecat.purchases.models.PurchaseType
+import com.revenuecat.purchases.models.StoreTransaction
+import org.json.JSONObject
+
+/**
+ * Code for handling in-app purchases and restorations directly by the application rather than by RevenueCat.
+ * These methods are called by a RevenueCat Paywall only when `Purchases.purchasesAreCompletedBy` is `MY_APP`.
+ *
+ * @property performPurchase Performs an in-app purchase.
+ * @property performRestore Perform a purchase restore for the given customer.
+ */
+class MyAppPurchaseLogic(
+    val performPurchase: suspend ((Activity, Package) -> MyAppPurchaseResult),
+    val performRestore: suspend ((CustomerInfo) -> MyAppRestoreResult),
+)
+
+/**
+ * Represents the result of a purchase attempt made by custom app-based code (not RevenueCat).
+ */
+sealed class MyAppPurchaseResult {
+    /**
+     * Indicates a successful purchase.
+     *
+     * @property purchase The purchase details if available. This object is only used to call the
+     * `PaywallListener` `onPurchaseCompleted` callback. If this callback is not needed, it does not need to be
+     * provided, and the callback will not be called.
+     */
+    data class Success(val purchase: MyAppPurchase? = null) : MyAppPurchaseResult()
+
+    /**
+     * Indicates the purchase was cancelled.
+     */
+    object Cancellation : MyAppPurchaseResult()
+
+    /**
+     * Indicates an error occurred during the purchase attempt.
+     *
+     * @property error Details of the error that occurred.
+     */
+    data class Error(val error: PurchasesError) : MyAppPurchaseResult()
+}
+
+/**
+ * Represents the result of a restore purchases attempt.
+ */
+sealed class MyAppRestoreResult {
+    /**
+     * Indicates a successful restore operation.
+     */
+    object Success : MyAppRestoreResult()
+
+    /**
+     * Indicates an error occurred during the restore attempt.
+     *
+     * @property error Details of the error that occurred.
+     */
+    data class Error(val error: PurchasesError) : MyAppRestoreResult()
+}
+
+/**
+ * Represents a purchase made in the application.
+ *
+ * @property purchase The actual purchase object from the store.
+ * @property productType The type of product being purchased. Defaults to [ProductType.UNKNOWN].
+ * @property presentedOfferingContext Context of the offering presented to the user, if applicable.
+ * @property subscriptionOptionId The subscription option ID from Google, if applicable.
+ * @property replacementMode The replacement mode for Google subscriptions, if applicable.
+ */
+class MyAppPurchase(
+    val purchase: Purchase,
```

Comment:
This is only compatible with Google, not Amazon. However, I believe, we don't need this object, since our SDK should automatically pick purchases up and send them to our servers I believe. We do need to know whether the purchase was successful or not to display a success/error state in the paywall though. Lmk if I'm missing something though!

Reply:
Yeah, going with ~observer mode~ Purchases Completed By My App ™️ doesn't mean we would miss them, so we might be fine without having to do the MyAppPurchase dance. 

Slightly more nuance with Amazon in that we still need you to directly report the purchases to us. But that is kind of independent of this, right? We'd still require you to call the Amazon-specific method? 

cc @vegaro

Reply:
The _only_ reason for this is to enable calling the PaywallListener purchaseCompleted callback (as discussed/ing in Slack, at meeting).

Reply:
Fwiw for Google, you still need to call `syncPurchases` for us to catch the new purchase, for Amazon, you are required to call `syncAmazonPurchase` which is slightly different because `syncPurchases` doesn't work when using our SDK side by side your own implementation, due to the Amazon SDK not being able to handle two purchase listeners at the same time.

Reply:
> Fwiw for Google, you still need to call syncPurchases for us to catch the new purchase

Hmm wouldn't we get it in the listener ourselves too without calling `syncPurchases`?

For Amazon, yeah they need to call the corresponding sync method.